### PR TITLE
sql: initial support for regtype

### PR DESCRIPTION
**Pull request by mkysel**
This does not fully support `pgcli` as it does not support casting regtype to text, but it is a step forward.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR adds a release note for any
  [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

File: src/sql/src/plan/typeconv.rs [hunk 1]
```diff
@@ -251,6 +256,29 @@ lazy_static! {
                     )
                     END
             )"),
+            // A regtype represents a type by oid. Converting from string to regtype
+            // does a lookup of the function name and expects exactly one type to
+            // match it. You can also specify (in postgres) a string that's a valid
+            // int4 and it'll happily cast it (without verifying that the int4 matches
+            // a function oid). To support this, use a SQL expression that checks if
```
Comment (mjibson) [inline src/sql/src/plan/typeconv.rs:263]: "type oid"

File: test/sqllogictest/types.slt [hunk 2]
```diff
@@ -832,3 +832,76 @@ query T
...
+query T
+SELECT 'date'::regtype::oid
+----
+1082
+
+query error type "nonexistent" does not exist
+SELECT 'nonexistent'::regtype
+
+query T
+SELECT 'date'::regtype
```
Comment (mjibson) [inline test/sqllogictest/types.slt:888]: This is duplicated above.

**Review by mjibson — APPROVED** [review]
LGTM if you squish the commits.

**Comment by mkysel** [comment]
Given that `regproc` is not mentioned in our docs, I will follow the same pattern and not add a RN.

**Review by philip-stoev — APPROVED** [review]
Thank you for the tests! There is nothing I can add at this time.

I assume the final goal of this project is to be able to run the queries from https://github.com/dbcli/pgcli/blob/master/pgcli/pgexecute.py right? If so, once those queries become runnable, we should add them to our test suite.

**Pull request merged**